### ISSUE TITLE
Help w/ Functional Wrapping

### ISSUE THREAD

**Issue by a-lipson** [IBM/fp-go#47]
Hello Dr Carsten, 

I was working on a project, and I encountered a function that I wanted to try to wrap into a function into a IOEither type.

The function's signature looks like this:
```go
package docconv

func ConvertDocx(r io.Reader) (string, map[string]string, error)
```

`docconv.ConvertDocx :: io.Reader -> IOEither error (string, (string -> string))`

So, this is my attempt at a wrapper for the function. 

```go
// wrapper function
func convertDocx(r io.Reader) IOE.IOEither[error, T.Tuple2[string, map[string]string]] {

	text, meta, err := docconv.ConvertDocx(r)
	if err != nil {
		return IOE.Left[T.Tuple2[string, map[string]string]](err)
	}

	return IOE.Right[error](T.MakeTuple2(text, meta))
}
```

However, I was wondering if it's possible to more concisely convert a function with more than one output parameter into a function that returns a tuple of those parameters. I'd imagine this is very similar to the `TupledN` functions, but just for the function outputs.
Does this functionality exist in fp-go? If not, is there a reason why?

Of course, even if this functionality is available, functions that have more than one input and more than one output at the same time are still difficult to work with. 

To illustrate,

```go
func OutTupled3[F ~func(T) (R1, R2, R3), T, R1 R2, R3 any](f F) func(T) Tuple3[R1, R2, R3] {
    return func(t T) Tuple3[R1, R2, R3] {
        t1, t2, t3 := f(t)        
        return T.MakeTuple3[t1, t2, t3]
    }
}
```

The next part of the use case doesn't really make sense anywhere aside from functions that throw some arguments and then an error, where the either-ness is implicit in the function definition that all other outputs are going to be invalid (even if, perhaps they were initialized to some value aside from their zero value). 
A counterexample is perhaps an IO function with a couple steps, where side effects could occur even when the ultimate return value of the function is an error.

Nonetheless, now we must convert the tuple which, at this point, contains an error as it's last element, and convert it into a IOEither that contains the error as the left argument, and then the remaining size N-1 tuple as the right argument. 
I suppose these functions could be named `EitherizeN` or `IOEitherizeN`?
The choice of where to put the selected field is not particularly consequential because there are `Swap` functions for the Either types.

---

The second question I have is related to converting between `Either[A, B]` and `Option[B]`. My specific case is with, `IOEither[error, []byte]` and `Option[error]`.

I'm trying to create a function which takes a string path and a bytes slice content using `ioeither/file.WriteFile(path, os.ModePerm)` and the function in creates, and then returns an `Option[error]` in the case that the IOEither[error, []byte] had a value in the left side. 

I have tried both the `ioeither.Fold` and `ioeither.GetOrElse` methods and am currently looking at your examples in context/readerioeither and either/examples_extract_test to see if I can figure it out. 

What I've landed on thus far is an `io.IO[error]`, which I don't think I can get out of unless the original type was Either[error, IO[[]byte]].

```go
err := F.Pipe3(
	content, 
	IOF.WriteFile(path, os.ModePerm),
	IOE.Swap, 
	IOE.GetOrElse(F.Constant1[[]byte](IO.Of[error](nil))),
)
```

So, my question is, how might you solve this problem. My hope is that the solution is rather elegant and I'm overcomplicating the problem by having taken an erroneous approach from the get-go.

Thank you very much once again!

**Comment by a-lipson** [comment]
I was taking a look at 
https://github.com/IBM/fp-go/blob/main/ioeither/examples_extract_test.go
and saw that you can extract the value of an IO via what seemed calling the value as a function.
https://github.com/IBM/fp-go/blob/34844bcfc26750b9196987e8d03bf6d641d685ff/ioeither/examples_extract_test.go#L32-L34

I was wondering if there exists a function or some composition of function that achieves the following.
```go
func IOEval[A any](a IO[A]) A { return a() }
```

Or, if such a function would not be ideal because we want to hold onto the fact that a value came from an IO operation up until we need to use the value?

Here's the example that I'm currently working with

```go
func TestGetContentFromDocx(t *testing.T) {

	eval := func(b IO.IO[[]byte]) []byte { return b() }

	out := F.Pipe5(
		"../data/document.docx",
		getContentFromPath, // returns IOEither[error, []byte]
		IOE.GetOrElse(F.Constant1[error](IO.Of([]byte{}))),
		eval,
		getTextFromDocx, // returns IOEither[error, string]
		IOE.GetOrElse(F.Constant1[error](IO.Of(""))),
	)

	fmt.Println(out())
}
```
I am under the impression that this composition of functions is not the idea way to produce the desired output, because there are the `GetOrElse` and evaluation calls. I could remove the evaluation calls by redefining the function `getTextFromDocx` function to take an `IO[[]byte]` instead of just a `[]byte`... but I don't think that's good form either.

This may be a case of trying to learn and work with functional concepts in golang before learning them in a more suitable environment first. As I mentioned in another issue post, I am learning Haskell as well, so perhaps I will be able to answer my own question soon.

Nonetheless, once again, thank you for your time!

**Comment by CarstenLeue** [comment]
About the second question:

> My specific case is with, IOEither[error, []byte] and Option[error].

I think the goal should be instead to convert from `IOEither[error, []byte]` to `IOOption[error]` since `IOEither` is an effectful operation and you need to convert this to another effectul monad. In your case to the `Option` equivalent which is `IOOption`.

One way to do this is:

```go

import (
	"fmt"
	"os"
	"path/filepath"
	"testing"

	F "github.com/IBM/fp-go/function"
	IOE "github.com/IBM/fp-go/ioeither"
	IOEF "github.com/IBM/fp-go/ioeither/file"
	IOEG "github.com/IBM/fp-go/ioeither/generic"
	IOO "github.com/IBM/fp-go/iooption"
)

	resIOO := F.Pipe2(
		content,
		IOEF.WriteFile(filepath.Join(tmpDir, "test.txt"), os.ModePerm),
		IOEG.Fold[IOE.IOEither[error, []byte]](
			IOO.Of[error],
			F.Ignore1of1[[]byte](IOO.None[error]),
		),
	)

```

It's not really nice since you have to use the generic version of the `Fold` function which forces you to specify the type. Reason is that in go `IO[Option]` is not the same (and also not compatible to ) as `IOOption` (although `type IOOption[A any] I.IO[O.Option[A]]`)

The other - better - option is this:

```go
	resIOO := F.Pipe3(
		content,
		IOEF.WriteFile(filepath.Join(tmpDir, "test.txt"), os.ModePerm),
		IOE.Swap[error, []byte],
		IOO.FromIOEither[[]byte, error],
	)
```

but I noticed that I failed to implement [IOO.FromIOEither], so you could not have found this solution. I have added the missing function, now.

**Comment by a-lipson** [comment]
Thank you so much for your assistance! 

I realized at some point that IOOption was the appropriate type after attempting to determine the nesting order IO[Option] or Option[IO]. Since preservation of IO is most important, and in the latter case, one could have a None which losses the effectual status, vs. IO[None], which retains the status.

With regards to the `IOO.FromIOEither`, I implemented a bad/io-unsafe version by using the non-generic `IOE.Fold` and "evaluating" the resulting `IO[Option]`s before casting them to `IOOptions` via `IOO.FromOption`.M
Anyways, thank you for implementing `FromIOEither`, I will now use that instead!

---

Additionally, thank you for providing a suggested conversion for the `convertDocx` and for noting the more modular approach of using a reader instead of some bytes directly.

I'm going to close this issue now.
Much appreciated!
